File: app/src/main/java/com/example/asthmaapplication/main/homepage/RegistrationFragment.java
package com.example.asthmaapplication.main.homepage;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;

import com.example.asthmaapplication.R;
import com.example.asthmaapplication.databinding.FragmentRegistrationBinding;
import com.example.asthmaapplication.main.common.BaseFragment;
import com.example.asthmaapplication.main.mainpage.MainActivity;
import com.example.asthmaapplication.main.utils.UIUtils;

import javax.annotation.Nullable;
import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import io.reactivex.annotations.NonNull;

@AndroidEntryPoint
public class RegistrationFragment extends BaseFragment {
    FragmentRegistrationBinding binding;
    HomePageViewModel viewModel;
    FragmentManager manager;
    FragmentTransaction transaction;
    TextWatcher watcher;
    SharedPreferences preferences;

    @Inject
    public RegistrationFragment() {

    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewModel = new ViewModelProvider(this).get(HomePageViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentRegistrationBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@Nullable View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        getPreferences();
        setActionBarTitle();

        binding.actionRegister.setEnabled(false);

        manager = getFragmentManager();
        transaction = manager.beginTransaction();

        UIUtils.addUnderlineFlag(binding.actionLogin);
        binding.actionLogin.setOnClickListener(v -> {
            launchLoginPage();
        });

        watcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                boolean allowRegister = (
                        !TextUtils.isEmpty(binding.emailLogin.getText().toString()) &&
                                !TextUtils.isEmpty(binding.passwordLogin.getText().toString()));

                binding.actionRegister.setEnabled(allowRegister);
                binding.actionRegister.setOnClickListener(v -> {
                    viewModel.register(binding.emailLogin.getText().toString(),
                            binding.passwordLogin.getText().toString());
                    resetRegistrationPage();
                });
            }
        };


        binding.emailLogin.addTextChangedListener(watcher);
        binding.passwordLogin.addTextChangedListener(watcher);

    }

    @Override
    public void onResume() {
        super.onResume();
    }

    public void getPreferences() {
        if (getContext() != null) {
            preferences = getContext().getSharedPreferences("user.prefs", Context.MODE_PRIVATE);
        }
    }

    public void resetRegistrationPage() {
        binding.emailLogin.setText(null);
        binding.passwordLogin.setText(null);
        getFragmentManager().popBackStackImmediate();
    }

    public void launchLoginPage() {
        LoginFragment fragment = new LoginFragment();
        transaction
                .addToBackStack(null)
                .replace(R.id.fragment_container, fragment)
                .commit();
    }

    public void setActionBarTitle() {
        setActionBarTitle(getString(R.string.registration_page_title));
    }

    @Override
    public View getRoot() {
        return binding.getRoot();
    }
}

File: app/src/main/java/com/example/asthmaapplication/main/homepage/RegistrationFragment.kt
package com.example.asthmaapplication.main.homepage

import dagger.hilt.android.AndroidEntryPoint
import com.example.asthmaapplication.main.common.BaseFragment
import android.text.TextWatcher
import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import com.example.asthmaapplication.main.utils.UIUtils
import android.text.Editable
import android.text.TextUtils
import android.view.View
import androidx.fragment.app.viewModels
import com.example.asthmaapplication.R
import com.example.asthmaapplication.databinding.FragmentRegistrationBinding
import com.example.asthmaapplication.main.common.BaseActivity
import com.example.asthmaapplication.main.common.Event
import com.example.asthmaapplication.main.utils.ActivityUtils

@AndroidEntryPoint
class RegistrationFragment: BaseFragment() {
    private lateinit var _binding: FragmentRegistrationBinding
    private lateinit var watcher: TextWatcher

    private val _viewModel: HomePageViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentRegistrationBinding.inflate(inflater, container, false)
        return _binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setActionBarTitle()
        (activity as BaseActivity?)?.showBackButton(true)
        UIUtils.addUnderlineFlag(_binding.actionLogin)

        _viewModel.loading.observe(viewLifecycleOwner) { setupLoading(it) }
        _viewModel.snackBar.observe(viewLifecycleOwner) { showSnackBar(it) }
        _viewModel.registrationSuccess.observe(viewLifecycleOwner) { setupRegistrationListener(it) }

        _binding.actionRegister.isEnabled = false
        _binding.actionLogin.setOnClickListener { launchLoginPage() }

        setupListeners()

    }

    private fun setupLoading(event: Event<Boolean>) {
        val result = event.contentIfNotHandled
        result?.let {
            _binding.progressIndicator.visibility = if (it) View.VISIBLE else View.GONE
        }
    }

    private fun setupRegistrationListener(event: Event<Boolean>) {
        event.contentIfNotHandled?.let {
            if (it) {
                fragmentManager!!.popBackStackImmediate()
            }
        }
    }

    private fun setupListeners() {
        watcher = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                val allowRegister = !TextUtils.isEmpty(_binding.emailLogin.text.toString()) &&
                        !TextUtils.isEmpty(_binding.passwordLogin.text.toString())
                _binding.actionRegister.isEnabled = allowRegister
                _binding.actionRegister.setOnClickListener {
                    _viewModel.register(
                        _binding.emailLogin.text.toString(),
                        _binding.passwordLogin.text.toString()
                    )
                    resetRegistrationPage()
                }
            }
        }
        _binding.emailLogin.addTextChangedListener(watcher)
        _binding.passwordLogin.addTextChangedListener(watcher)
    }

    private fun resetRegistrationPage() {
        _binding.emailLogin.text = null
        _binding.passwordLogin.text = null
    }

    private fun launchLoginPage() {
        val fragment = LoginFragment()
        ActivityUtils.addFragmentWithBackStack(
            parentFragmentManager,
            fragment,
            R.id.fragment_container,
            null
        )
    }

    fun setActionBarTitle() {
        setActionBarTitle(getString(R.string.registration_page_title))
    }

    override fun getRoot(): View {
        return _binding.root
    }
}